server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -366,7 +371,69 @@ private String getLocalSegmentFilename(String remoteFilename) {
+     * we just need to read the latest metadata file. All the stale metadata files can be safely deleted.
+     * @param lastNMetadataFilesToKeep number of metadata files to keep
+     * @throws IOException in case of I/O error while reading from / writing to remote segment store
+     */
+    public void deleteStaleSegments(int lastNMetadataFilesToKeep) throws IOException {
+        Collection<String> metadataFiles = remoteMetadataDirectory.listFilesByPrefix(MetadataFilenameUtils.METADATA_PREFIX);
+        List<String> sortedMetadataFileList = metadataFiles.stream().sorted(METADATA_FILENAME_COMPARATOR).collect(Collectors.toList());
+        if (sortedMetadataFileList.size() <= lastNMetadataFilesToKeep) {
+            logger.info(
+                "Number of commits in remote segment store={}, lastNMetadataFilesToKeep={}",
+                sortedMetadataFileList.size(),
+                lastNMetadataFilesToKeep
+            );
+            return;
+        }
+        List<String> latestNMetadataFiles = sortedMetadataFileList.subList(
+            sortedMetadataFileList.size() - lastNMetadataFilesToKeep,
+            sortedMetadataFileList.size()
+        );
+        Map<String, UploadedSegmentMetadata> activeSegmentFilesMetadataMap = new HashMap<>();
+        Set<String> activeSegmentRemoteFilenames = new HashSet<>();
+        for (String metadataFile : latestNMetadataFiles) {
+            Map<String, UploadedSegmentMetadata> segmentMetadataMap = readMetadataFile(metadataFile);
+            activeSegmentFilesMetadataMap.putAll(segmentMetadataMap);
+            activeSegmentRemoteFilenames.addAll(
+                segmentMetadataMap.values().stream().map(metadata -> metadata.uploadedFilename).collect(Collectors.toSet())
+            );
+        }
+        for (String metadataFile : sortedMetadataFileList.subList(0, sortedMetadataFileList.size() - lastNMetadataFilesToKeep)) {
+            Map<String, UploadedSegmentMetadata> staleSegmentFilesMetadataMap = readMetadataFile(metadataFile);
+            Set<String> staleSegmentRemoteFilenames = staleSegmentFilesMetadataMap.values()
+                .stream()
+                .map(metadata -> metadata.uploadedFilename)
+                .collect(Collectors.toSet());
+            AtomicBoolean deletionSuccessful = new AtomicBoolean(true);
+            staleSegmentRemoteFilenames.stream().filter(file -> !activeSegmentRemoteFilenames.contains(file)).forEach(file -> {
+                try {
+                    remoteDataDirectory.deleteFile(file);
+                    if (!activeSegmentFilesMetadataMap.containsKey(getLocalSegmentFilename(file))) {
+                        segmentsUploadedToRemoteStore.remove(getLocalSegmentFilename(file));
+                    }
+                } catch (NoSuchFileException e) {
+                    logger.info("Segment file {} corresponding to metadata file {} does not exist in remote", file, metadataFile);
```

Comment:
The underlying blobstore for remote store supports bulk deletes, do you think we could add a `deleteFiles` interface to leverage the same?

Reply:
Yes, we can do that but that involves deviating from directory API that currently supports only single file delete. I am trying to keep the interface same as much as possible (making changes only if there is no other way). 

Let me think more on this. I guess we can have a common solution for bulk upload (referring to your comment [here](https://github.com/opensearch-project/OpenSearch/pull/4240#discussion_r956366622)) and delete. 
I have created a tracking issue here: https://github.com/opensearch-project/OpenSearch/issues/4317